### DISCUSSION TITLE
A story of a v9 migration that is blocked

### DISCUSSION THREAD

**Discussion by jgarplind** [eslint/eslint#18453] [General]
As per discussions with @nzakas on BlueSky, I'm adding a discussion here to document the issues we are facing with upgrading to ESLint v9.

Our code base har origins from `create-react-app`, and has since moved on to use Vite.

We have a `.eslintrc.json` file which looks like so:

```
{
  "extends": [
    "react-app",
    "prettier",
    "plugin:jsx-a11y/recommended",
    "plugin:@tanstack/eslint-plugin-query/recommended"
  ],
  "plugins": ["jsx-a11y", "@tanstack/query"]
}
```

making use of the following packages: `eslint-config-prettier`, `eslint-config-react-app`, `@tanstack/eslint-plugin-query`. `jsx-a11y` is included in `eslint-config-react-app`.

When naively upgrading to ESLint v9, the first error message is this:

> Oops! Something went wrong! :(
ESLint: 9.2.0
Error: Could not find config file.

This is expected, so I tried to make a basic `eslint.config.js` file, which looks like this:

```
export default [
  {
    files: ["**/*.ts", "**/*.tsx"],
  },
];
```

Obviously, we now have TypeScript issues (e.g. `error  Parsing error: Unexpected token :`), since we need a plugin to support TS. This used to be part of `react-app`. Naively using the package like so:

```
import reactApp from 'eslint-config-react-app'

export default [
  reactApp,
  {
    files: ["**/*.ts", "**/*.tsx"],
  },
];
```

gives another error

> Oops! Something went wrong! :(
ESLint: 9.2.0
A config object is using the "extends" key, which is not supported in flat config system.
Instead of "extends", you can include config objects that you'd like to extend from directly in the flat config array.
Please see the following page for more information:
https://eslint.org/docs/latest/use/configure/migration-guide#predefined-and-shareable-configs

which I attribute to the fact that the package (`eslint-config-react-app`) does not yet support v9 (as per https://github.com/jsx-eslint/eslint-plugin-react/issues/3699).

Further down the line, similar issues are likely to arise with `@tanstack/eslint-plugin-query` (as per https://github.com/TanStack/query/pull/7253).

All things considered, I think we are stuck with ESLint 8 for now, unless we want to make a major revamp of our linting setup.

---

I think this is a somewhat common experience in the ESLint community right now, and a major reason why the v9 migration has not been very smooth.

Another thing to note is that our upgrade is driven by dependabot. Had it not been for dependabot, we may have postponed upgrading for quite some time, and not run into these issues before they were solved by the package maintainers.

**Comment by nzakas** [comment]
Thanks for sharing these details. The first thing I notice is that the error message when the config file wasn't found isn't very helpful. We can definitely do something about that.

Out of curiosity, did you follow the link that was mentioned in the error message? The goal was to get you to this section: https://eslint.org/docs/latest/use/configure/migration-guide#using-eslintrc-configs-in-flat-config

Though admittedly, that is a bit further down the page so we probably need to fix that link too.

**Reply by jgarplind** [reply]
> Out of curiosity, did you follow the link that was mentioned in the error message? The goal was to get you to this section: https://eslint.org/docs/latest/use/configure/migration-guide#using-eslintrc-configs-in-flat-config

I probably did at some point (since I recognize the content), but I'm not sure if I found it through other means or by following that link.

I also recall being a bit scared off by this code block:
```
import { FlatCompat } from "@eslint/eslintrc";
import path from "path";
import { fileURLToPath } from "url";

// mimic CommonJS variables -- not needed if using CommonJS
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const compat = new FlatCompat({
    baseDirectory: __dirname
});

export default [

    // mimic ESLintRC-style extends
    ...compat.extends("eslint-config-my-config"),
];
```

but looking at it now it looks fairly reasonable.

So while writing this response, I tried this:

```
import { FlatCompat } from "@eslint/eslintrc";
import path from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const compat = new FlatCompat({
  baseDirectory: __dirname,
});

export default [
  ...compat.extends("eslint-config-react-app"),
  {
    files: ["**/*.ts", "**/*.tsx"],
  },
];
```

But there is just another error message behind the corner, and I'm not even sure what `rushstack` has to do with anything and whether it is my package manager, ESLint, `eslint-config-react-app` or something else that is acting up. I'm on deep water and this is where I would just revert to v8 and wait it out.

> Oops! Something went wrong! :(
ESLint: 9.2.0
Error: Cannot read config file: C:\Users\x\x\x\x\node_modules\.pnpm\eslint-config-react-app@7.0.1_@babel+plugin-syntax-flow@7.24.1_@babel+plugin-transform-react-_d7bhogsngpszor2bxne77n6nru\node_modules\eslint-config-react-app\base.js
Error: The ESLint patch script has only been tested with ESLint version 6.x, 7.x, and 8.x. (Your version: 9.2.0)
Consider reporting a GitHub issue:
https://github.com/microsoft/rushstack/issues

**Reply by nzakas** [reply]
Oh wow. So it looks like `react-app` is using the [Rush Stack ESLint Patch](https://github.com/microsoft/rushstack/tree/main/eslint/eslint-patch) in their [config file](https://github.com/facebook/create-react-app/blob/0a827f69ab0d2ee3871ba9b71350031d8a81b7ae/packages/eslint-config-react-app/base.js#L10-L11).

So, this is basically a patch written by some folks at Microsoft that changes the way config resolution inside of ESLint works by monkey-patching things (an objectively awful thing to do).

I would say report an issue at the URL mentioned. I'm guessing they aren't actually doing anything that won't work with ESLint v9, they're just covering their bases.

**Reply by jgarplind** [reply]
> I would say report an issue at the URL mentioned.

I would, but I'm on water so deep I don't even know how to formulate such an issue, and I don't care enough about using ESLint v9 today that I will try to get involved further even further up the stream of dependencies.

Sounds a bit harsh, but it is the reality that there is only so much time and effort I will spend to get tooling working.

**Reply by nzakas** [reply]
Fair enough. It looks like there's already an issue:
https://github.com/microsoft/rushstack/issues/4635

**Reply by nzakas** [reply]
Working on a PR: https://github.com/microsoft/rushstack/pull/4719

**Comment by jgarplind** [comment]
FWIW, a small update:

I noticed that `eslint-plugin-react` now has [support for v9](https://github.com/jsx-eslint/eslint-plugin-react/releases/tag/v7.35.0) so I decided to give upgrading another shot.

Untangling things, our direct dependency is on https://www.npmjs.com/package/eslint-config-react-app, which along with the rest of Create React App appears to be unmaintained. Hence, there is little hope for that particular config to be updated.

In the meanwhile, I gave re-creating a similar setup a feeble attempt, but was promptly blocked by [another plugin not supporting v9 yet](https://github.com/facebook/react/issues/28313).

Not interested in diving deep into canary releases, plugin peer compatibility and whatnot, I will once more pause this effort, waiting for the v9-ecosystem to mature a bit.

**Reply by nzakas** [reply]
Have you tried the [config migrator](https://eslint.org/blog/2024/05/eslint-configuration-migrator/)? We're far enough along that it might just work automatically for you.

**Reply by jgarplind** [reply]
Using the migration tool, our v8 config (see OP) became this:

```
import jsxA11Y from "eslint-plugin-jsx-a11y";
import tanstack/query from "@tanstack/eslint-plugin-query";
import path from "node:path";
import { fileURLToPath } from "node:url";
import js from "@eslint/js";
import { FlatCompat } from "@eslint/eslintrc";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const compat = new FlatCompat({
  baseDirectory: __dirname,
  recommendedConfig: js.configs.recommended,
  allConfig: js.configs.all,
});

export default [
  {
    ignores: [],
  },
  ...compat.extends(
    "react-app",
    "prettier",
    "plugin:jsx-a11y/recommended",
    "plugin:@tanstack/eslint-plugin-query/recommended",
  ),
  {
    plugins: {
      "jsx-a11y": jsxA11Y,
        "@tanstack/query": tanstack/query,
    },
  },
];
```

Other than the obvious syntax error (`tanstack/query`, which I changed to `tanstackQuery`), the output looks reasonable.

Running this config with v9.9, the first error was:

> ConfigError: Config (unnamed): Key "plugins": Cannot redefine plugin "@tanstack/query".

I don't fully understand this, but am curious to see if there is more, so I commented out that row under `plugins` and tried again.

Then, I got a different error, once again related to the problematic `eslint-config-react-app` plugin('s dependencies):

> TypeError: Error while loading rule 'flowtype/define-flow-type': context.getAllComments is not a function

We don't use Flow, so we have no use for that rule.

Therefore, I went into `node_modules` and removed it manually, just to see what would happen.

Next, this happened:

> TypeError: context.getScope is not a function
Occurred while linting ...\index.tsx:1
Rule: "@typescript-eslint/no-redeclare"

Checking the source code using the tag for the package version referred by the now close to ancient `eslint-config-react-app`, specifically https://github.com/typescript-eslint/typescript-eslint/blob/v5.62.0/packages/eslint-plugin/src/rules/no-redeclare.ts, we can see that the syntax used there is `context.getScope()`, unlike the current source code: https://github.com/typescript-eslint/typescript-eslint/blob/e83466bce5ecb173d786e6ffef10729a760a754d/packages/eslint-plugin/src/rules/no-redeclare.ts which uses the probably up-to-date `context.sourceCode.getScope`.

Hacking away at this level is not something I feel will end up in a good place, so again, I feel we are probably forced (for better) to drop the unmaintained https://www.npmjs.com/package/eslint-config-react-app.

The migration path is not clear to me, but that is obviously not up to you but rather the (now missing) plugin authors to define.

**Reply by jgarplind** [reply]
I did a re-take and went to https://vitejs.dev/guide/#scaffolding-your-first-vite-project and saw how they did it.

Then I started visiting every dependency of `eslint-config-react-app` and installed them myself.

Short of 1-2 plugins (https://github.com/import-js/eslint-plugin-import/issues/2948#issuecomment-2148832701 stands out in a bad way), it was somewhat easy to get things running.

My new config which is not quite identical looks like this, littered with comments since each plugin seems to do things slightly differently:

```
// Generated using `npm create vite@latest my-app -- --template react-ts` 2024-08-14

import js from "@eslint/js";
import globals from "globals";
import react from "eslint-plugin-react";
import reactHooks from "eslint-plugin-react-hooks";
import reactRefresh from "eslint-plugin-react-refresh";
import tseslint from "typescript-eslint";
import jsxA11Y from "eslint-plugin-jsx-a11y";
import tanstackQuery from "@tanstack/eslint-plugin-query";
import jest from "eslint-plugin-jest";
import testingLibrary from "eslint-plugin-testing-library";

// Consider adding `eslint-plugin-import` when https://github.com/import-js/eslint-plugin-import/issues/2948#issuecomment-2148832701 is resolved.

// Required for unmaintained plugins to have v9 compat, see https://eslint.org/blog/2024/05/eslint-compatibility-utilities/
import { fixupConfigRules } from "@eslint/compat";

export default tseslint.config({
  // https://eslint.org/docs/latest/use/configure/configuration-files#configuring-shared-settings
  // as required by https://github.com/jsx-eslint/eslint-plugin-react/?tab=readme-ov-file#configuration-legacy-eslintrc-
  settings: {
    react: {
      version: "detect",
    },
  },
  extends: [
    js.configs.recommended,
    ...tseslint.configs.recommended,
    // https://github.com/jsx-eslint/eslint-plugin-react?tab=readme-ov-file#flat-configs
    react.configs.flat.recommended,
    react.configs.flat["jsx-runtime"],
    // https://tanstack.com/query/latest/docs/eslint/eslint-plugin-query#recommended-setup
    ...tanstackQuery.configs["flat/recommended"],
    // https://github.com/jsx-eslint/eslint-plugin-jsx-a11y?tab=readme-ov-file#usage---flat-config-eslintconfigjs
    jsxA11Y.flatConfigs.recommended,
    // https://github.com/jest-community/eslint-plugin-jest?tab=readme-ov-file#recommended
    // https://github.com/jest-community/eslint-plugin-jest?tab=readme-ov-file#running-rules-only-on-test-related-files
    {
      files: ["**/__tests__/**/*.[jt]s?(x)", "**/?(*.)+(spec|test).[jt]s?(x)"],
      ...jest.configs["flat/recommended"],
      rules: {
        ...jest.configs["flat/recommended"].rules,
        // Error: Error while loading rule 'jest/no-deprecated-functions': Unable to detect Jest version - please ensure jest package is installed, or otherwise set version explicitly
        "jest/no-deprecated-functions": "off",
      },
    },
    // https://github.com/testing-library/eslint-plugin-testing-library?tab=readme-ov-file#react
    // https://github.com/testing-library/eslint-plugin-testing-library?tab=readme-ov-file#run-the-plugin-only-against-test-files
    {
      files: ["**/__tests__/**/*.[jt]s?(x)", "**/?(*.)+(spec|test).[jt]s?(x)"],
      // Until v9 ready, see:
      // https://github.com/testing-library/eslint-plugin-testing-library/issues/899
      // https://github.com/testing-library/eslint-plugin-testing-library/issues/924
      ...fixupConfigRules(testingLibrary.configs["flat/react"])[0],
    },
  ],
  files: ["**/*.{ts,tsx}"],
  ignores: ["dist"],
  languageOptions: {
    ecmaVersion: 2020,
    globals: globals.browser,
  },
  plugins: {
    "react-hooks": reactHooks,
    "react-refresh": reactRefresh,
  },
  rules: {
    ...reactHooks.configs.recommended.rules,
    "react-refresh/only-export-components": [
      "warn",
      { allowConstantExport: true },
    ],
  },
});
```

I think this will most likely do for now. I have 157 new lint issues (probably due to updated versions of dependencies, and some different rules enabled) to address, but most likely a setup that is more future proof and not terribly complex.

**Reply by nzakas** [reply]
I'm glad you got something working, and thanks for sharing the issues with the config migrator. There are definite bugs there that we can fix, and knowing that `eslint-config-react-app` is a problem will allow us to wrap that for others.
